controlplane/kubeadm/controllers/controller.go

```diff
@@ -154,16 +154,27 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
 		return ctrl.Result{Requeue: true}, nil
 	}
 
-	// Add finalizer first if not exist to avoid the race condition between init and delete
-	if !controllerutil.ContainsFinalizer(kcp, controlplanev1.KubeadmControlPlaneFinalizer) {
-		controllerutil.AddFinalizer(kcp, controlplanev1.KubeadmControlPlaneFinalizer)
-
+	defer func() {
+		// Always attempt to Patch the KubeadmControlPlane object and status after each reconciliation.
 		if err := patchHelper.Patch(ctx, kcp); err != nil {
-			logger.Error(err, "Failed to patch KubeadmControlPlane to add finalizer")
-			return ctrl.Result{}, err
+			logger.Error(err, "Failed to patch KubeadmControlPlane")
+			reterr = kerrors.NewAggregate([]error{reterr, err})
 		}
 
-		return ctrl.Result{Requeue: true}, nil
+		// TODO: remove this as soon as we have a proper remote cluster cache in place.
+		// Make KCP to requeue in case status is not ready, so we can check for node status without waiting for a full resync (by default 10 minutes).
+		// Only requeue if we are not going in exponential backoff due to error, or if we are not already re-queueing, or if the object has a deletion timestamp.
+		if reterr == nil && !res.Requeue && !(res.RequeueAfter > 0) && kcp.ObjectMeta.DeletionTimestamp.IsZero() {
+			if !kcp.Status.Ready {
+				res = ctrl.Result{RequeueAfter: 20 * time.Second}
+			}
+		}
```

Comment:
Can we add this to the other defer block so that we don't set `RequeueAfter` when we are setting the finalizer?